modules/apps/frontend-js/frontend-js-importmaps-extender/src/main/java/com/liferay/frontend/js/importmaps/extender/internal/servlet/taglib/JSImportMapsCache.java

```diff
@@ -0,0 +1,187 @@
+/**
+ * SPDX-FileCopyrightText: (c) 2024 Liferay, Inc. https://liferay.com
+ * SPDX-License-Identifier: LGPL-2.1-or-later OR LicenseRef-Liferay-DXP-EULA-2.0.0-2023-06
+ */
+
+package com.liferay.frontend.js.importmaps.extender.internal.servlet.taglib;
+
+import com.liferay.portal.kernel.json.JSONFactory;
+import com.liferay.portal.kernel.json.JSONObject;
+import com.liferay.portal.kernel.json.JSONUtil;
+import com.liferay.portal.kernel.service.CompanyLocalService;
+
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * @author Iván Zaera Avellón
+ */
+public class JSImportMapsCache {
+
+	public static final long COMPANY_ID_ALL = 0;
+
+	public JSImportMapsCache(
+		CompanyLocalService companyLocalService, JSONFactory jsonFactory) {
+
+		_companyLocalService = companyLocalService;
+		_jsonFactory = jsonFactory;
+	}
+
+	public synchronized String getImportMaps(long companyId) {
+		if (!_valid) {
+			_rebuildImportMaps();
+
+			_valid = true;
+		}
+
+		return _importMapsMap.get(companyId);
+	}
+
+	public synchronized void invalidate() {
+		_valid = false;
+	}
+
+	public synchronized JSImportMapsRegistration register(
+		long companyId, JSONObject jsonObject, String scope) {
+
+		if (scope == null) {
+			Map<Long, JSONObject> globalImportMapsJSONObjects =
+				_getGlobalImportMapsJSONObjects(companyId);
+
+			long globalId = _nextGlobalId++;
+
+			globalImportMapsJSONObjects.put(globalId, jsonObject);
+
+			invalidate();
+
+			return () -> globalImportMapsJSONObjects.remove(globalId);
+		}
+
+		Map<String, JSONObject> scopedImportMapsJSONObjects =
+			_getScopedImportMapsJSONObjects(companyId);
+
+		scopedImportMapsJSONObjects.put(scope, jsonObject);
+
+		invalidate();
+
+		return () -> scopedImportMapsJSONObjects.remove(scope);
```

Comment:
Here I see 2 issues:

- This deletion operation is invoked from `ServiceTrackerCustomizer.removedService()` and that makes me believe it might not be `syncrhonized`
- Also, this call is not invoking `invalidate()` so import maps remain as they are after the registration is removed

Reply:
Fixed :tada: -> https://github.com/liferay-frontend/liferay-portal/pull/4579/commits/0bbc4b91df8ef1db78d646f35552fec4b09ca0e3 and https://github.com/liferay-frontend/liferay-portal/pull/4579/commits/0897c6f77af207de90ac3628590efced29f7004e